### PR TITLE
Inserted check for PyObject_IsInstance in THPVariableCheck 

### LINKED ISSUE #65084: THPVariableCheck should check the return code for PyObject_IsInstance
## 🐛 Bug

Otherwise this can lead to hard-to-debug failures if it returns -1.
https://github.com/pytorch/pytorch/blob/54d060a8c91c27167925b380698331850352c841/torch/csrc/autograd/python_variable.h#L42-L45

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

### PR DESCRIPTION
**Pull request by francescocastelli**
Summary: Inserted check for the return of PyObject_IsInstance to capture the case in which it raises an exception and return -1. When this happen THPVariable_Check now throws a python_error to signal the exception.

Fixes #65084

### CODE REVIEW

File: torch/csrc/autograd/python_variable.h [hunk 1]
```diff
@@ -43,7 +44,10 @@ static inline bool THPVariable_CheckExact(PyObject *obj) {
 
 inline bool THPVariable_Check(PyObject *obj)
 {
-  return THPVariableClass && PyObject_IsInstance(obj, THPVariableClass);
+  const auto result = PyObject_IsInstance(obj, THPVariableClass);
+  if (result == -1)
+      throw python_error();
+  return THPVariableClass && result;
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.h:50]: The `&&` here to guard the isinstance check and not give it an invalid `THPVariableClass`. This should be preserved and this function should juste return false (without running the isinstance check at all) if `THPVariableClass` is nullptr.

**Review by albanD — COMMENTED** [review]
Thanks for fixing this. Just a small update.

File: torch/csrc/autograd/python_variable.h (hunk 1 above)
Comment (francescocastelli) [reply torch/csrc/autograd/python_variable.h]: @albanD Yes, right. Just fixed!

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**